kura/org.eclipse.kura.core.crypto/src/main/java/org/eclipse/kura/core/crypto/CryptoServiceImpl.java

```diff
@@ -89,21 +120,50 @@ public char[] encryptAes(char[] value) throws KuraException {
             byte[] iv = new byte[IV_SIZE];
             this.random.nextBytes(iv);
             c.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(AUTH_TAG_LENGTH_BIT, iv));
-            byte[] encryptedBytes = c.doFinal(new String(value).getBytes());
-
+            byte[] encryptedBytes = c.doFinal(charArrayToByteArray(value));
             String ivString = base64Encode(iv);
             String encryptedMessage = base64Encode(encryptedBytes);
+
             return (ivString + ENCRYPTED_STRING_SEPARATOR + encryptedMessage).toCharArray();
         } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
             throw new KuraException(KuraErrorCode.OPERATION_NOT_SUPPORTED, "encrypt");
-        } catch (InvalidKeyException | IllegalBlockSizeException | BadPaddingException e) {
+        } catch (InvalidKeyException | IllegalBlockSizeException | BadPaddingException | CharacterCodingException e) {
             throw new KuraException(KuraErrorCode.ENCODE_ERROR, VALUE_EXCEPTION_CAUSE);
         } catch (InvalidAlgorithmParameterException e) {
             throw new KuraException(KuraErrorCode.ENCODE_ERROR, PARAMETER_EXCEPTION_CAUSE);
         }
 
     }
 
+    private byte[] charArrayToByteArray(char[] value) throws CharacterCodingException {
+
+        ByteBuffer byteBuffer;
+        try {
+            byteBuffer = this.utf8Encoder.encode(CharBuffer.wrap(value));
+        } catch (CharacterCodingException e) {
+            // fallback for backward compatibility
+            byteBuffer = this.platformEncoder.encode(CharBuffer.wrap(value));
```

Comment:
Does the behaviour here change compared to the past?
The encoder is set to CodingErrorAction.REPORT so it fails on issues on encoding.
But in the current code it acts silently replacing characters. So, to keep backward compatibility, we should use the old code

Reply:
I'll change the CodingErrorAction with replace for the platform encoder/decoder